scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -471,7 +471,18 @@ trait TypedPipe[+T] extends Serializable {
    * Reasonably common shortcut for cases of total associative/commutative reduction
    * returns a ValuePipe with only one element if there is any input, otherwise EmptyValue.
    */
-  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = ComputedValue(groupAll.sum[U].values)
+  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = {
+    // every 1000 items, compact.
+    lazy implicit val batchedSG = Batched.compactingSemigroup[U](1000)
+    ComputedValue(map { t => ((), Batched[U](t)) }
+      .sumByLocalKeys
+      // remove the Batched before going to the reducers
+      .map { case (_, batched) => batched.sum }
```

Comment:
(Deleted my previous comments because I wasn't thinking straight)

It would not be difficult, I think, to write a single-key version of `AdaptiveCache` which used the same trick to tune the batch-size for `Batched` to make sure that we never ran out of memory.

It's also interesting to think about how this optimization could be built, more generally, into `AdaptiveCache` itself - that is, it could try to grow the `Batched` size as well as the number of keys in the cache. Once it starts hitting memory pressure it can decide whether to tradeoff reducing the batch size to grow the number of keys further.